File: nim_core_v2/android/src/main/kotlin/com/netease/nimflutter/services/FLTLoginService.kt
/*
 * Copyright (c) 2022 NetEase, Inc. All rights reserved.
 * Use of this source code is governed by a MIT license that can be
 * found in the LICENSE file.
 */

package com.netease.nimflutter.services

import android.content.Context
import com.netease.nimflutter.FLTConstant.paramErrorCode
import com.netease.nimflutter.FLTService
import com.netease.nimflutter.MethodChannelSuspendResult
import com.netease.nimflutter.NimCore
import com.netease.nimflutter.NimResult
import com.netease.nimflutter.nimProviderTimeout
import com.netease.nimflutter.toLoginOptions
import com.netease.nimflutter.toMap
import com.netease.nimflutter.toNIMLoginClient
import com.netease.nimlib.sdk.NIMClient
import com.netease.nimlib.sdk.v2.V2NIMError
import com.netease.nimlib.sdk.v2.auth.V2NIMLoginDetailListener
import com.netease.nimlib.sdk.v2.auth.V2NIMLoginExtensionProvider
import com.netease.nimlib.sdk.v2.auth.V2NIMLoginListener
import com.netease.nimlib.sdk.v2.auth.V2NIMLoginService
import com.netease.nimlib.sdk.v2.auth.V2NIMTokenProvider
import com.netease.nimlib.sdk.v2.auth.enums.V2NIMConnectStatus
import com.netease.nimlib.sdk.v2.auth.enums.V2NIMDataSyncState
import com.netease.nimlib.sdk.v2.auth.enums.V2NIMDataSyncType
import com.netease.nimlib.sdk.v2.auth.enums.V2NIMLoginClientChange
import com.netease.nimlib.sdk.v2.auth.enums.V2NIMLoginStatus
import com.netease.nimlib.sdk.v2.auth.model.V2NIMKickedOfflineDetail
import com.netease.nimlib.sdk.v2.auth.model.V2NIMLoginClient
import com.netease.nimlib.sdk.v2.auth.option.V2NIMLoginOption
import com.netease.yunxin.kit.alog.ALog
import kotlin.coroutines.resume
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.channels.onFailure
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withTimeoutOrNull

/**
 * flutter 登录组件
 *
 * @constructor
 *
 * @param applicationContext
 * @param nimCore
 */
class FLTLoginService(
    applicationContext: Context,
    nimCore: NimCore
) : FLTService(applicationContext, nimCore) {

    override val serviceName: String = "LoginService"

    init {
        nimCore.onInitialized {
            loginListener()
            loginDetailListener()
            registerFlutterMethodCalls(
                "login" to this::login,
                "logout" to this::logout,
                "getLoginUser" to this::getLoginUser,
                "getLoginStatus" to this::getLoginStatus,
                "getLoginClients" to this::getLoginClients,
                "kickOffline" to this::kickOffline,
                "getKickedOfflineDetail" to this::getKickedOfflineDetail,
                "getConnectStatus" to this::getConnectStatus,
                "getDataSync" to this::getDataSync,
                "getChatroomLinkAddress" to this::getChatroomLinkAddress,
                "setReconnectDelayProvider" to this::setReconnectDelayProvider
            )
        }
    }

    @ExperimentalCoroutinesApi
    private fun loginListener() {
        callbackFlow<Pair<String, Map<String, Any?>?>> {
            val listener = object : V2NIMLoginListener {
                override fun onLoginStatus(status: V2NIMLoginStatus?) {
                    ALog.i(serviceName, "onLoginStatus: $status")
                    trySend(
                        Pair<String, Map<String, Any?>>(
                            "onLoginStatus",
                            mapOf(
                                "status" to
                                    status?.value
                            )
                        )
                    ).onFailure {
                        ALog.i(serviceName, "send login status fail: ${it?.message}")
                    }
                }

                override fun onLoginFailed(error: V2NIMError?) {
                    ALog.i(serviceName, "onLoginFailed: $error")
                    trySend(Pair<String, Map<String, Any?>?>("onLoginFailed", error?.toMap())).onFailure {
                        ALog.i(serviceName, "send login failed fail: ${it?.message}")
                    }
                }

                override fun onKickedOffline(detail: V2NIMKickedOfflineDetail?) {
                    ALog.i(serviceName, "onKickedOffline: $detail")
                    trySend(Pair<String, Map<String, Any?>?>("onKickedOffline", detail?.toMap())).onFailure {
                        ALog.i(serviceName, "send kicked offline fail: ${it?.message}")
                    }
                }

                override fun onLoginClientChanged(
                    change: V2NIMLoginClientChange?,
                    clients: MutableList<V2NIMLoginClient>?
                ) {
                    ALog.i(serviceName, "onLoginClientsChange: $change")
                    trySend(
                        Pair<String, Map<String, Any?>?>(
                            "onLoginClientChanged",
                            mapOf(
                                "change" to change?.value,
                                "clients" to clients?.map { it.toMap() }?.toList()
                            )
                        )
                    ).onFailure {
                        ALog.i(serviceName, "send login client changed fail: ${it?.message}")
                    }
                }
            }
            NIMClient.getService(V2NIMLoginService::class.java).apply {
                this.addLoginListener(listener)
                awaitClose {
                    this.removeLoginListener(listener)
                }
            }
        }.onEach { event ->
            notifyEvent(
                method = event.first,
                arguments = event.second as Map<String, Any?>
            )
        }.launchIn(nimCore.lifeCycleScope)
    }

    @ExperimentalCoroutinesApi
    private fun loginDetailListener() {
        callbackFlow<Pair<String, Map<String, Any?>?>> {
            val listener = object : V2NIMLoginDetailListener {
                override fun onConnectStatus(status: V2NIMConnectStatus?) {
                    ALog.i(serviceName, "onConnectStatus: $status")
                    trySend(
                        Pair<String, Map<String, Any?>>(
                            "onConnectStatus",
                            mapOf(
                                "status" to
                                    status?.value
                            )
                        )
                    ).onFailure {
                        ALog.i(serviceName, "send connect status fail: ${it?.message}")
                    }
                }

                override fun onDisconnected(error: V2NIMError?) {
                    ALog.i(serviceName, "onDisconnected: $error")
                    trySend(Pair<String, Map<String, Any?>?>("onDisconnected", error?.toMap())).onFailure {
                        ALog.i(serviceName, "send disconnected fail: ${it?.message}")
                    }
                }

                override fun onConnectFailed(error: V2NIMError?) {
                    ALog.i(serviceName, "onConnectFailed: $error")
                    trySend(Pair<String, Map<String, Any?>?>("onConnectFailed", error?.toMap())).onFailure {
                        ALog.i(serviceName, "send connect failed fail: ${it?.message}")
                    }
                }

                override fun onDataSync(
                    type: V2NIMDataSyncType?,
                    state: V2NIMDataSyncState?,
                    error: V2NIMError?
                ) {
                    ALog.i(serviceName, "onDataSync: $type")
                    trySend(
                        Pair<String, Map<String, Any?>?>(
                            "onDataSync",
                            mapOf(
                                "type" to if (type != null) type.value else null,
                                "state" to if (state != null) state.value else null,
                                "error" to error?.toMap()
                            )
                        )
                    ).onFailure {
                        ALog.i(serviceName, "send data sync fail: ${it?.message}")
                    }
                }
            }
            NIMClient.getService(V2NIMLoginService::class.java).apply {
                this.addLoginDetailListener(listener)
                awaitClose {
                    this.removeLoginDetailListener(listener)
                }
            }
        }.onEach { event ->
            notifyEvent(
                method = event.first,
                arguments = event.second as Map<String, Any?>
            )
        }.launchIn(nimCore.lifeCycleScope)
    }

    private val loginExtensionProvider =
        V2NIMLoginExtensionProvider { accountId ->
            if (accountId != null) {
                runBlocking {
                    withTimeoutOrNull(nimProviderTimeout) {
                        suspendCancellableCoroutine<Any?> { continuation ->
                            notifyEvent(
                                method = "getLoginExtension",
                                arguments = mapOf("accountId" to accountId),
                                callback = MethodChannelSuspendResult(continuation)
                            )
                        } as? String?
                    }
                }
            } else {
                null
            }
        }

    private val loginTokenProvider = V2NIMTokenProvider { accountId ->
        if (accountId != null) {
            runBlocking {
                withTimeoutOrNull(nimProviderTimeout) {
                    suspendCancellableCoroutine<Any?> { continuation ->
                        notifyEvent(
                            method = "getToken",
                            arguments = mapOf("accountId" to accountId),
                            callback = MethodChannelSuspendResult(continuation)
                        )
                    } as? String?
                }
            }
        } else {
            null
        }
    }

    private suspend fun login(arguments: Map<String, *>): NimResult<Void> {
        val accountId = arguments["accountId"] as String?
        val token = arguments["token"] as String?
        if (accountId?.isEmpty() == true || token?.isEmpty() == true) {
            return NimResult(code = paramErrorCode, errorDetails = "accountId or token is empty")
        }
        return suspendCancellableCoroutine { cont ->
            val optionMap = arguments["option"] as Map<String, *>?
            val option = optionMap?.toLoginOptions() ?: V2NIMLoginOption()
            option.tokenProvider = loginTokenProvider
            option.loginExtensionProvider = loginExtensionProvider
            NIMClient.getService(V2NIMLoginService::class.java).login(
                accountId,
                token,
                option,
                {
                    cont.resume(NimResult(0, data = it))
                },
                {
                    cont.resume(NimResult(it.code, errorDetails = it.desc))
                }
            )
        }
    }

    private suspend fun logout(arguments: Map<String, *>): NimResult<Void> {
        return suspendCancellableCoroutine { cont ->
            NIMClient.getService(V2NIMLoginService::class.java).logout(
                {
                    cont.resume(NimResult(0, data = it))
                },
                {
                    cont.resume(NimResult(it.code, errorDetails = it.desc))
                }
            )
        }
    }

    private suspend fun getLoginUser(arguments: Map<String, *>): NimResult<String?> {
        return suspendCancellableCoroutine { cont ->
            NIMClient.getService(V2NIMLoginService::class.java).loginUser.let {
                cont.resume(NimResult(0, data = it))
            }
        }
    }

    private suspend fun getLoginStatus(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            NIMClient.getService(V2NIMLoginService::class.java).loginStatus.let {
                cont.resume(
                    NimResult(
                        0,
                        data = mapOf<String, Any>(
                            "status" to it.value
                        )
                    )
                )
            }
        }
    }

    private suspend fun getLoginClients(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            NIMClient.getService(V2NIMLoginService::class.java).loginClients.let { clients ->
                cont.resume(
                    NimResult(
                        0,
                        data = mapOf<String, Any>(
                            "loginClient" to clients.map { it.toMap() }.toList()
                        )
                    )
                )
            }
        }
    }

    private suspend fun kickOffline(arguments: Map<String, *>): NimResult<Void> {
        val clientMap = arguments["client"] as Map<String, *>?
        val client = clientMap?.toNIMLoginClient()
            ?: return NimResult(
                paramErrorCode,
                errorDetails = "client is null"
            )
        return suspendCancellableCoroutine { cont ->
            NIMClient.getService(V2NIMLoginService::class.java).kickOffline(
                client,
                {
                    cont.resume(NimResult(0, data = it))
                },
                {
                    cont.resume(NimResult(it.code, errorDetails = it.desc))
                }
            )
        }
    }

    private suspend fun getKickedOfflineDetail(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            val detail = NIMClient.getService(V2NIMLoginService::class.java).kickedOfflineDetail
            cont.resume(NimResult(0, data = detail?.toMap()))
        }
    }

    private suspend fun getConnectStatus(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            val status = NIMClient.getService(V2NIMLoginService::class.java).connectStatus
            cont.resume(NimResult(0, data = mapOf("status" to status.value)))
        }
    }

    private suspend fun getDataSync(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        return suspendCancellableCoroutine { cont ->
            val dataSync = NIMClient.getService(V2NIMLoginService::class.java).dataSync
            cont.resume(
                NimResult(
                    0,
                    data = mapOf(
                        "dataSync" to dataSync?.map {
                            it.toMap()
                        }?.toList()
                    )
                )
            )
        }
    }

    private suspend fun getChatroomLinkAddress(arguments: Map<String, *>): NimResult<Map<String, Any?>?> {
        val roomId = arguments["roomId"] as String?
        if (roomId?.isEmpty() == true) {
            return NimResult(
                paramErrorCode,
                errorDetails = "roomId is empty"
            )
        }
        return suspendCancellableCoroutine { cont ->
            NIMClient.getService(V2NIMLoginService::class.java).getChatroomLinkAddress(
                roomId,
                {
                    cont.resume(NimResult(0, data = mapOf("linkAddress" to it)))
                },
                {
                    cont.resume(NimResult(it.code, errorDetails = it.desc))
                }
            )
        }
    }

    private suspend fun setReconnectDelayProvider(arguments: Map<String, *>): NimResult<Void> {
        NIMClient.getService(V2NIMLoginService::class.java).setReconnectDelayProvider {
            runBlocking {
                withTimeoutOrNull(nimProviderTimeout) {
                    val a = suspendCancellableCoroutine<Any?> { continuation ->
                        notifyEvent(
                            method = "getReconnectDelay",
                            arguments = mapOf("delay" to it),
                            callback = MethodChannelSuspendResult(continuation)
                        )
                    } as? Int?
                    a?.toLong()
                } ?: 0
            }
        }
        return NimResult(0)
    }
}

File: nim_core_v2/ios/Classes/services/FLTLoginService.swift
// Copyright (c) 2022 NetEase, Inc. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

import Foundation
import NIMSDK

@objc
public protocol NEIMKitClientListener: NSObjectProtocol {
  @objc optional func onDataSync(_ type: V2NIMDataSyncType, state: V2NIMDataSyncState, error: V2NIMError?)
  /// 登录连接状态回调
  /// - Parameter status: 连接状态
  @objc optional func onConnectStatus(_ status: V2NIMConnectStatus)

  ///  连接失败回调
  ///  - Parameter error: 错误信息
  @objc optional func onConnectFailed(_ error: V2NIMError?)

  /// 断开连接回调
  /// - Parameter error: 错误信息
  @objc optional func onDisconnected(_ error: V2NIMError?)

  ///  登录状态变更回调
  ///  - Parameter status: 登录状态
  @objc optional func onLoginStatus(_ status: V2NIMLoginStatus)

  /// 登录失败回调
  /// - Parameter error: 错误信息
  @objc optional func onLoginFailed(_ error: V2NIMError)

  /// 被踢下线回调
  /// - Parameter detail: 被踢下线的详细信息
  @objc optional func onKickedOffline(_ detail: V2NIMKickedOfflineDetail)

  /// 登录客户端变更回调
  /// - Parameter change: 多端登录变动事件
  /// - Parameter clients: 登录客户端信息
  @objc optional func onLoginClientChanged(_ change: V2NIMLoginClientChange, clients: [V2NIMLoginClient]?)
}

@objc protocol IMKitClientDetailListenerManagerListener: NSObjectProtocol {
  /// 数据同步回调
  /// - Parameter type: 数据同步类型
  /// - Parameter state: 同步状态
  /// - Parameter error： 错误信息
  @objc optional func onDataSync(_ type: V2NIMDataSyncType, state: V2NIMDataSyncState, error: V2NIMError?)

  /// 登录连接状态回调
  /// - Parameter status: 连接状态
  @objc optional func onConnectStatus(_ status: V2NIMConnectStatus)

  ///  连接失败回调
  ///  - Parameter error: 错误信息
  @objc optional func onConnectFailed(_ error: V2NIMError?)

  /// 断开连接回调
  /// - Parameter error: 错误信息
  @objc optional func onDisconnected(_ error: V2NIMError?)
}

class IMKitClientDetailListenerManager: NSObject, V2NIMLoginDetailListener {
  public static let instance = IMKitClientDetailListenerManager()
  weak var delegate: IMKitClientDetailListenerManagerListener?
  override private init() {
    super.init()
    NIMSDK.shared().v2LoginService.add(self)
  }

  deinit {
    NIMSDK.shared().v2LoginService.remove(self)
  }

  ///  数据完成同步完成回调
  ///  - Parameter type: 数据同步类型
  ///  - Parameter state: 数据同步状态
  ///  - Parameter error: 错误信息
  public func onDataSync(_ type: V2NIMDataSyncType, state: V2NIMDataSyncState, error: V2NIMError?) {
    delegate?.onDataSync?(type, state: state, error: error)
  }

  /// 登录连接状态回调
  /// - Parameter status: 连接状态
  public func onConnectStatus(_ status: V2NIMConnectStatus) {
    delegate?.onConnectStatus?(status)
  }

  ///  连接失败回调
  ///  - Parameter error: 错误信息
  public func onConnectFailed(_ error: V2NIMError?) {
    delegate?.onConnectFailed?(error)
  }

  /// 断开连接回调
  /// - Parameter error: 错误信息
  public func onDisconnected(_ error: V2NIMError?) {
    delegate?.onDisconnected?(error)
  }
}

class FLTLoginService: FLTBaseService, FLTService, V2NIMLoginListener, V2NIMTokenProvider, V2NIMLoginExtensionProvider, V2NIMReconnectDelayProvider, IMKitClientDetailListenerManagerListener {
  func onLoginStatus(_ status: V2NIMLoginStatus) {
    notifyEvent(serviceName(), "onLoginStatus", ["status": status.rawValue])
  }

  func onLoginFailed(_ error: V2NIMError) {
    notifyEvent(serviceName(), "onLoginFailed", error.toDictionary())
  }

  func onKickedOffline(_ detail: V2NIMKickedOfflineDetail) {
    notifyEvent(serviceName(), "onKickedOffline", detail.toDictionary())
  }

  func onLoginClientChanged(_ change: V2NIMLoginClientChange, clients: [V2NIMLoginClient]?) {
    notifyEvent(serviceName(), "onLoginClientChanged", ["change": change.rawValue,
                                                        "clients": clients?.map { client in
                                                          client.toDictionary()
                                                        }])
  }

  func onConnectFailed(_ error: V2NIMError?) {
    notifyEvent(serviceName(), "onConnectFailed", error?.toDictionary())
  }

  func onConnectStatus(_ status: V2NIMConnectStatus) {
    notifyEvent(serviceName(), "onConnectStatus", ["status": status.rawValue])
  }

  func onDisconnected(_ error: V2NIMError?) {
    notifyEvent(serviceName(), "onDisconnected", error?.toDictionary())
  }

  func onDataSync(_ type: V2NIMDataSyncType, state: V2NIMDataSyncState, error: V2NIMError?) {
    notifyEvent(serviceName(), "onDataSync", ["type": type.rawValue, "state": state.rawValue, "error": error?.toDictionary()])
  }

  func getReconnectDelay(_ delay: Int32) -> Int32 {
    let semaphore = DispatchSemaphore(value: 0)
    var customDelay = -1
    notifyEvent(
      serviceName(),
      "getReconnectDelay",
      ["delay": Int(delay) as Any],
      result: { r in
        if let value = r as? NSNumber {
          customDelay = value.intValue
        }
        semaphore.signal()
      }
    )
//    let timeout = DispatchTime.now() + .seconds(30)
    semaphore.wait()
    print("getReconnectDelay ", Int32(customDelay))
    return Int32(customDelay)
  }

  func getLoginExtension(_ accountId: String) -> String? {
    if !accountId.isEmpty {
      let semaphore = DispatchSemaphore(value: 0)
      var logExtension = ""
      notifyEvent(
        serviceName(),
        "getLoginExtension",
        ["accountId": accountId as Any],
        result: { r in
          logExtension = r as? String ?? ""
          semaphore.signal()
        }
      )
      semaphore.wait()
      print("getLoginExtension ", logExtension)
      return logExtension
    }
    return nil
  }

  func getToken(_ accountId: String) -> String? {
    if !accountId.isEmpty {
      let semaphore = DispatchSemaphore(value: 0)
      var token = ""
      notifyEvent(
        serviceName(),
        "getToken",
        ["accountId": accountId as Any],
        result: { r in
          token = r as? String ?? ""
          semaphore.signal()
        }
      )
      semaphore.wait()
      return token
    }
    return nil
  }

  private let paramErrorTip = "param Error"
  private let paramErrorCode = 199_414

  private func loginCallback(_ error: Error?, _ data: Any?, _ resultCallback: ResultCallback) {
    if let ns_error = error as NSError? {
      errorCallBack(resultCallback, ns_error.description, ns_error.code)
    } else {
      successCallBack(resultCallback, data)
    }
  }

  private var dynamicLoginTokens = [String: String]()

  override func onInitialized() {
    IMKitClientDetailListenerManager.instance.delegate = self
    NIMSDK.shared().v2LoginService.add(self)
  }

  deinit {
    NIMSDK.shared().v2LoginService.remove(self)
  }

  func serviceName() -> String {
    ServiceType.LoginService.rawValue
  }

  func onMethodCalled(_ method: String, _ arguments: [String: Any],
                      _ resultCallback: ResultCallback) {
    switch method {
    case "login":
      login(arguments, resultCallback)
    case "logout":
      logout(arguments, resultCallback)
    case "getLoginUser":
      getLoginUser(arguments, resultCallback)
    case "getLoginStatus":
      getLoginStatus(arguments, resultCallback)
    case "kickOffline":
      kickOffline(arguments, resultCallback)
    case "getKickedOfflineDetail":
      getKickedOfflineDetail(arguments, resultCallback)
    case "getConnectStatus":
      getConnectStatus(arguments, resultCallback)
    case "getDataSync":
      getDataSync(arguments, resultCallback)
    case "getChatroomLinkAddress":
      getChatroomLinkAddress(arguments, resultCallback)
    case "setReconnectDelayProvider":
      setReconnectDelayProvider(arguments, resultCallback)
    case "getLoginClients":
      getLoginClients(arguments, resultCallback)
    default:
      resultCallback.notImplemented()
    }
  }

  func register(_ nimCore: NimCore) {
    self.nimCore = nimCore
    nimCore.addService(self)
  }

  private func login(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let account = arguments["accountId"] as? String, let token = arguments["token"] as? String else {
      errorCallBack(resultCallback, paramErrorTip, paramErrorCode)
      return
    }

    DLog(s: "login argument : \(arguments)")
    var loginOption = V2NIMLoginOption()
    if let optionDic = arguments["option"] as? [String: Any?] {
      loginOption = V2NIMLoginOption.fromDic(optionDic)
    }

    loginOption.tokenProvider = self
    loginOption.loginExtensionProvider = self

    weak var weakSelf = self
    // 登录
    NIMSDK.shared().v2LoginService.login(account, token: token, option: loginOption) {
      weakSelf?.loginCallback(nil, nil, resultCallback)
    } failure: { error in
      weakSelf?.loginCallback(error.nserror, nil, resultCallback)
    }
  }

  private func logout(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    weak var weakSelf = self
    NIMSDK.shared().v2LoginService.logout {
      weakSelf?.loginCallback(nil, nil, resultCallback)
    } failure: { error in
      weakSelf?.loginCallback(error.nserror, nil, resultCallback)
    }
  }

  private func getLoginUser(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    let user = NIMSDK.shared().v2LoginService.getLoginUser()
    successCallBack(resultCallback, user)
  }

  private func getLoginStatus(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    let status = NIMSDK.shared().v2LoginService.getLoginStatus()
    successCallBack(resultCallback, ["status": status.rawValue])
  }

  private func kickOffline(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let clientDic = arguments["client"] as? [String: Any?] else {
      errorCallBack(resultCallback, paramErrorTip, paramErrorCode)
      return
    }
    let client = V2NIMLoginClient.fromDic(clientDic)
    weak var weakSelf = self
    NIMSDK.shared().v2LoginService.kickOffline(client) {
      weakSelf?.loginCallback(nil, nil, resultCallback)
    } failure: { error in
      weakSelf?.loginCallback(error.nserror, nil, resultCallback)
    }
  }

  private func getKickedOfflineDetail(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    let detail = NIMSDK.shared().v2LoginService.getKickedOfflineDetail()
    successCallBack(resultCallback, detail?.toDictionary())
  }

  private func getConnectStatus(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    let status = NIMSDK.shared().v2LoginService.getConnectStatus()
    successCallBack(resultCallback, ["status": status.rawValue])
  }

  private func getDataSync(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    let dataSync = NIMSDK.shared().v2LoginService.getDataSync()
    var list = [[String: Any?]]()
    dataSync?.forEach { detail in
      list.append(detail.toDictionary())
    }
    successCallBack(resultCallback,
                    ["dataSync": list])
  }

  private func getLoginClients(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    let loginClients = NIMSDK.shared().v2LoginService.getLoginClients()
    var list = [[String: Any?]]()
    loginClients?.forEach { client in
      list.append(client.toDictionary())
    }
    successCallBack(resultCallback, ["loginClient": list])
  }

  private func getChatroomLinkAddress(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
    guard let roomId = arguments["roomId"] as? String else {
      errorCallBack(resultCallback, paramErrorTip, paramErrorCode)
      return
    }
    weak var weakSelf = self
    NIMSDK.shared().v2LoginService.getChatroomLinkAddress(roomId) { address in
      weakSelf?.loginCallback(nil, address, resultCallback)
    } failure: { error in
      weakSelf?.loginCallback(error.nserror, nil, resultCallback)
    }
  }

  private func setReconnectDelayProvider(_ arguments: [String: Any], _ resultCallback: ResultCallback) {
//   NIMSDK.shared().v2LoginService.setReconnectDelayProvider(self)
    successCallBack(resultCallback, nil)
  }
}
